traits/trait_set_object.py

```diff
@@ -12,247 +12,425 @@
 
-    def remove(self, value):
-        set.remove(self, value)
+        Parameters
+        ----------
+        value : iterable
 
-        if self.name_items is not None:
-            self._send_trait_items_event(
-                self.name_items, TraitSetEvent(set([value]))
-            )
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                A set containing the removed values.
+            added : set
+                A set containing the added values.
+
+        """
+        values = set(value)
+        removed = self.intersection(values)
+        added = values.difference(removed)
+        if added:
+            added = self.validate(added)
+        added = added.difference(self)
+
+        if removed or added:
+            super().difference_update(removed)
+            super().update(added)
+            self.notify(removed, added)
 
     def discard(self, value):
+        """ Remove an element from a set if it is a member.
+
+        If the element is not a member, do nothing.
+
+        Parameters
+        ----------
+        value : any
+            An item in the set
+
+        Notes
+        -----
+        Parameters in the notification:
+            Fired if a value is removed.
+
+            removed : set
+                A set containing the removed values.
+            added : set
+                An empty set.
+
+        """
         if value in self:
             self.remove(value)
+            self.notify({value}, set())
 
     def pop(self):
-        value = set.pop(self)
+        """ Remove and return an arbitrary set element.
+        Raises KeyError if the set is empty.
+
+        Returns
+        -------
+        item : any
+            An element from the set.
+
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                A set containing the removed values.
+            added : set
+                An empty set.
 
-        if self.name_items is not None:
-            self._send_trait_items_event(
-                self.name_items, TraitSetEvent(set([value]))
-            )
-
-        return value
+        """
+        removed = super().pop()
+        self.notify({removed}, set())
+        return removed
 
     def clear(self):
-        removed = set(self)
-        set.clear(self)
+        """ Remove all elements from this set.
 
-        if self.name_items is not None:
-            self._send_trait_items_event(
-                self.name_items, TraitSetEvent(removed)
-            )
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                A set containing the removed values.
+            added : set
+                An empty set.
+
+        """
+        if not self:
+            return
+        removed = set(self)
+        super().clear()
+        self.notify(removed, set())
 
     def __reduce_ex__(self, protocol=None):
```

Comment:
Do we need to keep `__reduce_ex__`? It would be preferable to avoid it and just depend on `__getstate__`, `__setstate__`, `__getnewargs__`, etc., if possible.

Reply:
Moved to `TratSetObject` compatibility layer. But I couldn't find a good way to get rid of it.